{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "public List<SalesmanGenome> crossover(List<SalesmanGenome> parents) {\n",
    "    // Housekeeping\n",
    "    Random random = new Random();\n",
    "    int breakpoint = random.nextInt(genomeSize);\n",
    "    List<SalesmanGenome> children = new ArrayList<>();\n",
    "\n",
    "    // Copy parental genomes - we copy so we wouldn't modify in case they were\n",
    "    // chosen to participate in crossover multiple times\n",
    "    List<Integer> parent1Genome = new ArrayList<>(parents.get(0).getGenome());\n",
    "    List<Integer> parent2Genome = new ArrayList<>(parents.get(1).getGenome());\n",
    "    \n",
    "    // Creating child 1\n",
    "    for (int i = 0; i < breakpoint; i++) {\n",
    "        int newVal;\n",
    "        newVal = parent2Genome.get(i);\n",
    "        Collections.swap(parent1Genome, parent1Genome.indexOf(newVal), i);\n",
    "    }\n",
    "    children.add(new SalesmanGenome(parent1Genome, numberOfCities, travelPrices, startingCity));\n",
    "    parent1Genome = parents.get(0).getGenome(); // Reseting the edited parent\n",
    "    \n",
    "    // Creating child 2\n",
    "    for (int i = breakpoint; i < genomeSize; i++) {\n",
    "        int newVal = parent1Genome.get(i);\n",
    "        Collections.swap(parent2Genome, parent2Genome.indexOf(newVal), i);\n",
    "    }\n",
    "    children.add(new SalesmanGenome(parent2Genome, numberOfCities, travelPrices, startingCity));\n",
    "\n",
    "    return children;\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def partial_crossover(solution1, solution2):\n",
    "    parent1 = deepcopy(solution1.representation)\n",
    "    parent2 = deepcopy(solution2.representation)\n",
    "    \n",
    "    firstCrossPoint = np.random.randint(0,len(parent1)-2)\n",
    "    secondCrossPoint = np.random.randint(firstCrossPoint+1,len(parent1)-1)\n",
    "\n",
    "    #firstCrossPoint = 3\n",
    "    #secondCrossPoint = 6\n",
    "\n",
    "    print(firstCrossPoint, secondCrossPoint)\n",
    "\n",
    "    parent1MiddleCross = parent1[firstCrossPoint:secondCrossPoint]\n",
    "    parent2MiddleCross = parent2[firstCrossPoint:secondCrossPoint]\n",
    "\n",
    "    temp_child1 = parent1[:firstCrossPoint] + parent2MiddleCross + parent1[secondCrossPoint:]\n",
    "\n",
    "    temp_child2 = parent2[:firstCrossPoint] + parent1MiddleCross + parent2[secondCrossPoint:]\n",
    "    \n",
    "    \n",
    "    // Creating child 1\n",
    "    for (int i = 0; i < breakpoint; i++) {\n",
    "        int newVal;\n",
    "        newVal = parent2Genome.get(i);\n",
    "        Collections.swap(parent1Genome, parent1Genome.indexOf(newVal), i);\n",
    "    }\n",
    "    children.add(new SalesmanGenome(parent1Genome, numberOfCities, travelPrices, startingCity));\n",
    "    parent1Genome = parents.get(0).getGenome(); // Reseting the edited parent\n",
    "    \n",
    "    \n",
    "    for i in range(0, firstCrossPoint):\n",
    "        new_val = 0\n",
    "        new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def partially_mapped_crossover(problem, chromA, chromB, childIndex):\n",
    "    j = 0\n",
    "    item1 = 0\n",
    "    item2 = 0\n",
    "    pos1 = 0\n",
    "    pos2 = 0\n",
    "    thisChromo = Chromosome(self.mCityCount)\n",
    "    thisChromo = self.population[chromA]\n",
    "    thatChromo = Chromosome(self.mCityCount)\n",
    "    thatChromo = self.population[chromB]\n",
    "    \n",
    "    newChromo = Chromosome(self.mCityCount)\n",
    "    newChromo = self.population[childIndex]\n",
    "\n",
    "    crossPoint1 = random.randrange(self.mCityCount);\n",
    "    crossPoint2 = self.get_exclusive_random_integer(self.mCityCount, crossPoint1)\n",
    "    \n",
    "    if crossPoint2 < crossPoint1:\n",
    "        j = crossPoint1\n",
    "        crossPoint1 = crossPoint2\n",
    "        crossPoint2 = j\n",
    "\n",
    "    # Copy parentA genes to offspring.\n",
    "    for i in range(len(thisChromo)):\n",
    "        newChromo.set_data(i, thisChromo.get_data(i))\n",
    "\n",
    "    for i in range(crossPoint1, crossPoint2 + 1):\n",
    "        # Get the two items to swap.\n",
    "        item1 = thisChromo.get_data(i)\n",
    "        item2 = thatChromo.get_data(i)\n",
    "\n",
    "        # Get the items' positions in the offspring.\n",
    "        for k in range(self.mCityCount):\n",
    "            if newChromo.get_data(k) == item1:\n",
    "                pos1 = k\n",
    "            elif newChromo.get_data(k) == item2:\n",
    "                pos2 = k\n",
    "\n",
    "        # Swap them.\n",
    "        if item1 != item2:\n",
    "            newChromo.set_data(pos1, item2)\n",
    "            newChromo.set_data(pos2, item1)\n",
    "\n",
    "    return"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
